lib/yaml/constructor.py

```diff
@@ -10,16 +10,29 @@
 
 from .error import *
 from .nodes import *
+from .registry import RegistryMeta, ConstructorRegistry
 
+import threading
 import collections.abc, datetime, base64, binascii, re, sys, types
 
 class ConstructorError(MarkedYAMLError):
     pass
 
-class BaseConstructor:
 
-    yaml_constructors = {}
-    yaml_multi_constructors = {}
+class ConstructorSetup(threading.local):
+    def __init__(self):
+        self.initialized = False
+
+    def ensure_initialized(self):
+        if not self.initialized:
+            self.initialized = True
+            setup_constructors()
+
+
+class BaseConstructor(metaclass=RegistryMeta):
```

Comment:
Can we get an ASCII art diagram that illustrates the inheritance? Like the excalidraw diagram you showed me on our call. asciiflow.com is good for this.

Reply:
Done.